distributed/worker.py

```diff
@@ -3444,47 +3460,98 @@ async def find_missing(self) -> None:
+                        "which is not true.",
+                        self.name,
+                        ts,
+                    )
+                    instructions.append(
+                        MissingDataMsg(
+                            key=key, errant_worker=self.address, stimulus_id=stimulus_id
                         )
-                        # Do not mutate the input dict. That's rude
-                        workers = set(workers) - {self.address}
-                    dep_ts.who_has.update(workers)
+                    )
 
-                    for worker in workers:
-                        self.has_what[worker].add(dep)
-                        self.data_needed_per_worker[worker].push(dep_ts)
-        except Exception as e:  # pragma: no cover
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+            if ts.who_has == workers:
+                continue
 
-                pdb.set_trace()
-            raise
+            self.log.append(
+                (
+                    key,
+                    "update-who-has",
+                    list(ts.who_has),
+                    list(workers),
+                    stimulus_id,
+                    time(),
+                )
+            )
+
+            for worker in ts.who_has - workers:
+                self.has_what[worker].discard(key)
+                # Can't remove from self.data_needed_per_worker; there is logic
+                # in _select_keys_for_gather to deal with this
+
+            for worker in workers - ts.who_has:
+                self.has_what[worker].add(key)
+                if ts.state == "fetch":
+                    self.data_needed_per_worker[worker].push(ts)
+                    # All workers which previously held a replica of the key are either
+                    # in flight or busy. Kick off ensure_communicating to try fetching
+                    # the data from the new worker.
+                    ensure_communicating = True
+
+            ts.who_has = workers
+            # currently fetching -> can no longer be fetched -> transition to missing
+            # currently missing -> opportunity to be fetched -> transition to fetch
+            # any other state -> eventually, possibly, the task may transition to fetch
+            # or missing, at which point the relevant transitions will test who_has that
+            # we just updated. e.g. see the various transitions to fetch, which
+            # instead recommend transitioning to missing if who_has is empty.
+            if not workers and ts.state == "fetch":
+                recs[ts] = "missing"
+            elif workers and ts.state == "missing":
+                recs[ts] = "fetch"
+
+        if ensure_communicating:
+            recs, instructions = merge_recs_instructions(
+                (recs, instructions),
+                self._ensure_communicating(stimulus_id=stimulus_id),
```

Comment:
Like I said in https://github.com/dask/distributed/commit/cbccc86777789311ad5b6a6fe285eec8a5b63de2#r73797288, it seems odd to me to call `_ensure_communicating` directly. Why not just append an `EnsureCommunicatingAfterTransitions` instruction on L3535?

Reply:
Fair. We'll probably revert it soon as per our recent chat with @fjetter , but it makes sense to have an organic behaviour for now.